### PR TITLE
Encryption component

### PR DESCRIPTION
**Pull request by ccorcoveanu**
Added new component for `EncryptedData`.

It uses the same key derivation mechanism as the account keyfile (in fact, that was refactored to use `EncryptedData` under the hood which is more general-purpose).

Link to component description:
https://docs.google.com/document/d/1Xxb39JJkf8B9UzzQZd3KeOzL6nomoOxOyxKpxDKRgiE/edit#

### CODE REVIEW

File: src/walletcore/userWallet.ts [hunk 1]
```diff
@@ -85,14 +46,23 @@ export class UserWallet {
...
 
-    /**
-     * Will take about:
-     *  - 80-90 ms in Node.js, on a i3-8100 CPU @ 3.60GHz
-     *  - 350-360 ms in browser (Firefox), on a i3-8100 CPU @ 3.60GHz
-     */
-    private static generateDerivedKey(password: Buffer, salt: Buffer, kdfparams: ScryptKeyDerivationParams): Buffer {
-        const derivedKey = scryptsy(password, salt, kdfparams.n, kdfparams.r, kdfparams.p, kdfparams.dklen);
-        return derivedKey;
+    static ecFromJSON(keyfileObject: any): EncryptedData {
```
Comment (andreibancioiu) [inline src/walletcore/userWallet.ts:49]: What does the prefix `ec` mean?

File: src/crypto/encrypt.spec.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { assert } from "chai";
```
Comment (andreibancioiu) [inline src/crypto/encrypt.spec.ts:1]: 👍 for this test.

File: src/crypto/decryptor.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    const computedMAC = crypto.createHmac(DigestAlgorithm, derivedKeySecondHalf).update(ciphertext).digest();
+    const actualMAC = data.mac;
+
+    if (computedMAC.toString("hex") !== actualMAC) {
+      throw new errors.ErrWallet("MAC mismatch, possibly wrong password");
+    }
+
+    const decipher = crypto.createDecipheriv(data.cipher, derivedKeyFirstHalf, iv);
+
+    return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
```
Comment (andreibancioiu) [inline src/crypto/decryptor.ts:25]: So, the padding is still done in `userWallet.ts`.

I can confirm there's no change in logic - only refactoring ✔️

File: src/crypto/encryptor.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import crypto from "crypto";
+import { Randomness } from "./randomness";
+import { ScryptKeyDerivationParams } from "./derivationParams";
+import { CipherAlgorithm, DigestAlgorithm, Version, KeyDerivationFunction } from "./constants";
+import {EncryptedData} from "./encryptedData";
+
+export class Encryptor {
+  public static encrypt(data: Buffer, password: string, randomness: Randomness = new Randomness()): EncryptedData {
+    const kdParams = new ScryptKeyDerivationParams();
```
Comment (andreibancioiu) [inline src/crypto/encryptor.ts:9]: I confirm the login here is the one from the previous constructor of `UserWallet` ✔️

**Review by andreibancioiu — APPROVED** [review]
👍

File: src/walletcore/userWallet.ts (hunk 1 above)
Comment (ccorcoveanu) [reply src/walletcore/userWallet.ts]: It was a typo, should have been `ed` for "Encrypted Data"

**Review by andreibancioiu — APPROVED** [review]

**Pull request merged**